source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +62,108 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
+      auto effective_weight = host_weight * locality_weight;
+      weighted_sum += effective_weight;
+      effective_weights[host] = effective_weight;
+    }
+  }
+
+  // We can't do anything sensible with no hosts.
+  if (weighted_sum == 0) {
+    return;
+  }
 
-  uint64_t hashes_per_host = 1;
-  if (hosts.size() < min_ring_size) {
-    hashes_per_host = min_ring_size / hosts.size();
-    if ((min_ring_size % hosts.size()) != 0) {
-      hashes_per_host++;
+  // Determine the valid range of replication factors given the weighted sum and ring size bounds,
+  // and the actual replication factor within that range. This may differ from the configured
+  // replication factor if provided. Note that we use floating point math here and round up when
+  // adding hosts to the ring, rather than trying to preserve exact integer ratios. This is good
+  // enough, since the distribution of hosts around the ring wouldn't exactly match the given
+  // weights anyway, even with a good hash function.
+  const double min_replication_factor =
+      static_cast<double>(min_ring_size) / static_cast<double>(weighted_sum);
+  const double max_replication_factor =
+      static_cast<double>(max_ring_size) / static_cast<double>(weighted_sum);
+  double replication_factor;
+  if (configured_replication_factor) {
+    // If the configuration includes a replication factor, try to use it and warn if not possible.
+    replication_factor = static_cast<double>(configured_replication_factor.value());
+    if (replication_factor < min_replication_factor) {
+      ENVOY_LOG(warn, "ring hash: replication_factor ({}) too small for minimum_ring_size ({})",
+                replication_factor, min_ring_size);
+      replication_factor = min_replication_factor;
+    } else if (replication_factor > max_replication_factor) {
+      ENVOY_LOG(warn, "ring hash: replication_factor ({}) too large for maximum_ring_size ({})",
+                replication_factor, max_ring_size);
+      replication_factor = max_replication_factor;
     }
+  } else {
+    // If a replication factor isn't provided, use the old behavior for backward compatibility: size
+    // the ring as small as possible while maintaining integer weights.
+    replication_factor = std::ceil(min_replication_factor);
   }
 
-  ENVOY_LOG(info, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size, hashes_per_host);
-  ring_.reserve(hosts.size() * hashes_per_host);
+  // Adjust the ring entry weights by multiplying the entry weights with the replication factor,
+  // rounding up to avoid losing entries with weights less than 1, and sum these to determine the
+  // actual ring size.
+  // NOTE: This rounding changes the effective replication factor from what was determined above.
+  //       In some contrived cases, we could actually exceed the configured max ring size, but this
+  //       is probably better behavior than rounding down and losing entries.
```

Comment:
Did you change your thinking on this one? I thought we had discussed it might make sense to lose a host with weight 1 if the other host was 127 and we only had a ring size of 32.